engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ExtensionMethodResolutionTest.java

```diff
@@ -0,0 +1,136 @@
+        """;
+    testProjectCompilationFailure(src, methodsOverloadErrorMessageMatcher);
+  }
+
+  @Test
+  public void extensionMethodAndNormalMethodConflictInDifferentModules() throws IOException {
+    var modSrc = """
+        type T
+            foo x = x
+        """;
+    var mainSrc = """
+        from project.Mod import T
+        T.foo x y = x + y
+        """;
+    var projDir = createProject("Proj", mainSrc, tempFolder);
+    var modSrcFile = projDir.resolve("src").resolve("Mod.enso");
+    Files.writeString(modSrcFile, modSrc);
+    testProjectCompilationFailure(projDir, methodsOverloadErrorMessageMatcher);
+  }
+
+  @Test
+  public void extensionMethodAndNormalMethodConflictInThreeModules() throws IOException {
+    var tSrc = """
+        type T
+        """;
+    var modSrc = """
+        from project.T import T
+        T.foo x = x
+        """;
+    var mainSrc =
+        """
+        from project.T import T
+        # Make sure we import also the extension method from Mod
+        from project.Mod import all
+        T.foo x y = x + y
+        """;
+    var projDir = createProject("Proj", mainSrc, tempFolder);
+    var tSrcFile = projDir.resolve("src").resolve("T.enso");
+    Files.writeString(tSrcFile, tSrc);
+    var modSrcFile = projDir.resolve("src").resolve("Mod.enso");
+    Files.writeString(modSrcFile, modSrc);
+    testProjectCompilationFailure(projDir, methodsOverloadErrorMessageMatcher);
+  }
+
+  @Test
+  public void extensionMethodAndNormalMethodConflictInDifferentProjects() throws IOException {
+    var libSrc = """
+        type T
+            foo x = x
+        """;
+    createProject("Lib", libSrc, tempFolder);
+    var mainSrc = """
+        from local.Lib import T
+        T.foo x y = x + y
+        """;
+    var mainProjDir = createProject("Main", mainSrc, tempFolder);
+    testProjectCompilationFailure(mainProjDir, methodsOverloadErrorMessageMatcher);
+  }
+
+  private void testProjectCompilationFailure(String mainSrc, Matcher<String> errorMessageMatcher)
+      throws IOException {
+    var projDir = createProject("Proj", mainSrc, tempFolder);
+    testProjectCompilationFailure(projDir, errorMessageMatcher);
+  }
+
+  private void testProjectCompilationFailure(
+      Path mainProjDir, Matcher<String> errorMessageMatcher) {
+    var out = new ByteArrayOutputStream();
+    try (var ctx =
+        defaultContextBuilder()
+            .option(RuntimeOptions.PROJECT_ROOT, mainProjDir.toAbsolutePath().toString())
+            .option(RuntimeOptions.STRICT_ERRORS, "true")
+            .option(RuntimeOptions.DISABLE_IR_CACHES, "true")
+            .out(out)
+            .err(out)
+            .build()) {
+      var polyCtx = new PolyglotContext(ctx);
+      try {
+        polyCtx.getTopScope().compile(true);
+        fail("Expected compilation error: " + out);
+      } catch (PolyglotException e) {
+        assertThat(e.getMessage(), is(CompilationAbortedException.COMPILATION_ABORTED_MSG));
```

Comment:
Recognizing exceptions by their message? Unusual. I'd rather hardcode the message string directly into the test....


I think you can `assertTrue(e.isSyntax())`...

Reply:
Refactored to `assertTrue(e.isSyntaxError())` in 3cd30ab464b200946fe37690e28a0ee1ac6122ff